incubator/hnc/pkg/stats/stats.go

```diff
@@ -0,0 +1,187 @@
+package stats
+
+import (
+	"sync/atomic"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/config"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type counter int32
+
+type object struct {
+	gk              schema.GroupKind
+	totalReconciles counter
+	curReconciles   counter
+	apiWrites       counter
+}
+
+type objects map[schema.GroupKind]*object
+
+type stat struct {
+	// actionID is the number of controller actions devided by idles
+	actionID counter
+
+	// totalHierConfigReconciles is the total number of HierarchyConfig reconciliations.
+	totalHierConfigReconciles counter
+
+	// curHierConfigReconciles is the currently undergoing number of HierarchyConfig reconciliations.
+	curHierConfigReconciles counter
+
+	// hierConfigWrites is the total number of HierarchyConfig writes.
+	hierConfigWrites counter
+
+	// namespaceWrites is the total number of Namespace writes.
+	namespaceWrites counter
+
+	objects objects
+}
+
+var stats stat
+
+// StartHierConfigReconcile updates stats when hierarchyConfig
+// reconciliation starts.
+func StartHierConfigReconcile() {
+	stats.totalHierConfigReconciles.incr()
+	stats.curHierConfigReconciles.incr()
+}
+
+// StopHierConfigReconcile updates stats when hierarchyConfig
+// reconciliation finishes.
+func StopHierConfigReconcile() {
+	stats.curHierConfigReconciles.decr()
+}
+
+// StartObjReconcile updates the stats for objects with common GK
+// when an object reconciliation starts.
+func StartObjReconcile(gk schema.GroupKind) {
+	if stats.objects == nil {
+		stats.objects = make(map[schema.GroupKind]*object)
+	}
```

Comment:
Why is this needed? Don't we always call `initStats` before starting the controllers?

Reply:
I got error if I remove this part:
```
panic: assignment to entry in nil map [recovered]
        panic: assignment to entry in nil map
```
It's true that `stats.StartLoggingActivity()` is called before creating controllers. Maybe compiler cannot tell?

Reply:
That's a runtime error, not a compiler one. Which goroutine is generating that panic?

Reply:
Ah - if `testLog` is false, `stats.StartLoggingActivity()` is never called - maybe that's what you're seeing? If so, we can probably just say `if stats.objects == nil {return}` instead of creating it.

Reply:
ah right, exactly! ok will do that.

Reply:
This is done. I added a `testLogEnabled` flag and it's set to true in `stats.StartLoggingActivity()` func.

In each stats update func, it will check the flag first by:
```
if !testLogEnabled {
    return
}
```